packages/runtime/src/widgets/Form/TextInput.tsx

```diff
@@ -44,8 +53,102 @@ export const TextInput: React.FC<TextInputProps> = (props) => {
     }
   }, [value, formInstance, values?.id, values?.label]);
 
+  const inputType = useMemo(() => {
+    switch (values?.inputType) {
+      case "email":
+        return "email";
+      case "phone":
+        return "tel";
+      case "number":
+        return "number";
+      case "url":
+        return "url";
+      default:
+        return "text";
+    }
+  }, [values?.inputType]);
+
+  const patternValue = useMemo((): string | undefined | RegExp => {
+    if (!values?.mask) {
+      return;
+    }
+
+    // https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/pattern#overview
+    const special = [
+      "[",
+      "]",
+      "\\",
+      "/",
+      "^",
+      "$",
+      ".",
+      "|",
+      "?",
+      "*",
+      "+",
+      "(",
+      ")",
+      "{",
+      "}",
+    ];
+
+    let pattern = "^(?:";
+    for (const char of values.mask) {
+      switch (char) {
+        case "#":
+          pattern += "[0-9]"; // Match any digit
+          break;
+        case "A":
+          pattern += "[a-zA-Z]"; // Match any letter
+          break;
+        default:
+          if (special.includes(char)) {
+            pattern += `\\${char}`; // Escape special characters
+          } else {
+            pattern += char; // Include the character literally
+          }
+          break;
+      }
+    }
+    return pattern + ")$";
+  }, [values?.mask]);
+
+  const rules = useMemo(() => {
+    const rulesArray: Rule[] = [];
+
+    if (values?.validator?.minLength) {
+      rulesArray.push({
+        min: values.validator.minLength,
+        message: `The field  must be at least ${values.validator.minLength} characters long`,
+      });
+    }
+
+    if (values?.validator?.maxLength) {
+      rulesArray.push({
+        max: values.validator.maxLength,
+        message: `The field must be at most ${values.validator.maxLength} characters long`,
+      });
+    }
+
+    if (values?.validator?.regex) {
+      rulesArray.push({
+        pattern: new RegExp(values.validator.regex),
+        message: values.validator.regexError || "The field has invalid value",
+      });
+    }
+
+    if (values?.mask && patternValue) {
+      rulesArray.push({
+        pattern: new RegExp(patternValue),
+        message: `The field does not match the mask ${values.mask}`,
+      });
```

Comment:
Can we return a more user friendly value here? Something like `The value must be of the format <mask>`